astropy/utils/diff.py

```diff
@@ -181,7 +186,22 @@ def where_not_allclose(a, b, rtol=1e-5, atol=1e-8):
     if not np.all(np.isfinite(b)):
         b = np.ma.fix_invalid(b).data
 
+    absolute = np.abs(b - a)
+
     if atol == 0.0 and rtol == 0.0:
         # Use a faster comparison for the most simple (and common) case
-        return np.where(a != b)
-    return np.where(np.abs(a - b) > (atol + rtol * np.abs(b)))
+        indices = np.where(absolute > 0)
+    else:
+        indices = np.where(absolute > (atol + rtol * np.abs(b)))
```

Comment:
I replaced this else by these lines and now I get the expected results:
`else:

            finites = np.isfinite(absolute)

            with np.errstate(invalid="ignore", divide="ignore"):

                relative = absolute[finites] / np.abs(b[finites])

            max_absolute = float(np.max(absolute[finites]))

            max_relative = np.max(relative)
`